src/test/unit_tests/test_namespace_cache.js

```diff
@@ -0,0 +1,901 @@
+        const { obj } = await create_namespace_cache_and_read_obj({
+            recorder, size, ttl_ms, object_sdk, start, end });
+
+        await promise_utils.wait_until(() => {
+            try {
+                const cache_obj = recorder.get_obj('cache', obj.bucket, obj.key);
+                return !_.isUndefined(cache_obj) && cache_obj.num_parts === 0;
+            } catch (err) {
+                return false;
+            }
+        }, 2000, 100);
+    });
+
+    mocha.it('large range not cached', async () => {
+        const free = 100;
+        const object_sdk_large_range_read = {
+            should_run_triggers: () => null,
+            // Set free to small number
+            read_bucket_usage_info: () => ({ free }),
+        };
+
+        const size = block_size * 2;
+        const start = block_size - 100;
+        // end is exclusive
+        const end = start + free + 100;
+        const { obj } = await create_namespace_cache_and_read_obj({
+            recorder, size, ttl_ms, object_sdk: object_sdk_large_range_read, start, end });
+
+        P.delay(100);
+        try {
+            recorder.get_obj('cache', obj.bucket, obj.key);
+            assert(false);
+        } catch (err) {
+            assert(err.rpc_code === 'NO_SUCH_UPLOAD');
+        }
+    });
+
+    mocha.it('range read falls back to hub read if read from cache fails', async () => {
+        const size = block_size * 2;
+        let start = block_size - 2;
+        // end is exclusive
+        let end = start + 100;
+        const { ns_cache, obj, cache } = await create_namespace_cache_and_read_obj({
+            recorder, size, ttl_ms,
+            object_sdk: object_sdk,
+            start, end,
+        });
+        await promise_utils.wait_until(() => {
+            try {
+                const cache_obj = recorder.get_obj('cache', obj.bucket, obj.key);
+                return !_.isUndefined(cache_obj);
+            } catch (err) {
+                return false;
+            }
+        }, 2000, 100);
+
+        start = block_size;
+        end = block_size + 199;
+        const params = {
+            bucket: obj.bucket,
+            key: obj.key,
+            start,
+            end,
+        };
+
+        params.object_md = await ns_cache.read_object_md(params, object_sdk);
+        cache.set_trigger_err('read');
+        const err_trigger_time = new Date().getTime();
+        const stream = await ns_cache.read_object_stream(params, object_sdk);
+        cache.clear_trigger_err();
+        const read_buf = await buffer_utils.read_stream_join(stream);
+        const read_etag = crypto.createHash('md5').update(read_buf).digest('hex');
+        const expect_etag = crypto.createHash('md5').update(obj.buf.slice(start, end)).digest('hex');
+        assert(read_etag === expect_etag);
+
+        const read_hub_time = recorder.get_event('hub', obj.bucket, obj.key, EVENT_READ_OBJECT_STREAM);
+        assert(read_hub_time > err_trigger_time);
```

Comment:
Hi @paulxuzhang ,
In our Ci in the last weeks, we see that this line fails.
Is it need to be the operand `>` or if it can be `>=` ?
In our logs, I added a printing and we can see In case it fails:
> CHECK_VALUE: read_hub_time 1704205245518 err_trigger_time 1704205245518

cc: @Neon-White